# A very basic introduction to python, notebooks, data scraping with chatgpt, and some quick visualization/analysis

Hi all, first, I want to apologize for not making it; I was very much looking forward to seeing/meeting all of you. Second, I want to stress that *since I wasn't quite sure who all would be in attendance*, I wanted to make this a very open introduction.

Some of you may already know how to do all this (and much more), some of you may never have touched python before in your life. No worries either way. I'm going to try to keep this as self contained as possible. I'm sure there will be questions and I'm sure I'm going to make some mistakes trying to communicate all of this within the notebook itself; but, we're lucky enough to have the inimitable Ryan Burns here to help field those issues.

I'm going to try to explain everything you need to know here to be as inclusive and fun as possible. Please forgive any oversights.

If you enjoy this or you just want to ask some questions about whatever, you can reach me at thatchja@oregonstate.edu 

Here's what we're going to do:
1. Query a public data source
2. Geocode it.
3. Visualize it.
4. Compare that to ChatGPT results (not a direct comparison)

## First things first, what is this document and how do we interact with it?

This is a [jupyter notebook](https://jupyter.org/) which is the modern derivative of what was once an ipython notebook (this is why the file extension is .ipynb). At its core, it's a nicely open-source web application that lets us write code, execute it, and intersperse it with other media including (as I'm writing here) markdown.

I find it incredibly useful for distributing technical labs to students and have also used it for presentations and lectures. If you're locked into the ESRI ecosystem (no shame, some universities require it), you can actually run notebooks from within ArcGIS Pro as well. You can read more about that [here](https://www.esri.com/en-us/arcgis/products/arcgis-notebooks/overview).

But, right now, you're *probably* looking at this from where I've put it in github. Github will render the notebook, but it does so as a static image and doesn't execute the code (though it will keep visualizations that were present in the file when you commit it). I'm not going to assume folx know how to interact with github, but I'm also going to skip it as it's not **really** what I want to show you today. For what it's worth, you can get a desktop application for most OSes [here](https://desktop.github.com/), but I prefer to work with command-line git because it makes me feel like a hacker in a 90s movie (you can find documentation for that [here](https://docs.gitlab.com/ee/gitlab-basics/start-using-git.html)).

Something you might notice as we go on is that I'm going to provide a lot of links to documentation. That's because I'm constantly consulting it. I sincerely believe that one of the best skills one can have for working with code is the ability to parse documentation. Lame, but true. Also, I have a terrible memory.

So, if you're a github wizard, feel free to grab this however you want. If you're not, go to the repository ([here](https://github.com/jethatch/digiws)), click on that green button that says 'code' and download the zip file of everything you need.

Now, to be able to interact with it, we'll need to have Jupyter Notebooks installed. I am going to walk you through **one way** to do this using [Anaconda](https://www.anaconda.com/download/). There are lots of other options! You could use [miniconda](https://docs.conda.io/en/latest/miniconda.html), [mamba](https://mamba.readthedocs.io/en/latest/index.html), or even an online solution like [CoCalc](https://cocalc.com/features/jupyter-notebook).

Heck, you can actually use an end-around to do this with just ArcGIS Pro installed. For folx that end up teaching/working in labs that *don't* let you install libraries, you can usually end-around this restriction if ArcPro is there by using what ESRI calls the 'Python Command Prompt' which is ultimately their branding on the Anaconda prompt (ESRI likes to ingest open-source stuff into their systems and then slap their names on it. Ask Dillon Mahmoudi about this and tell him I told you to, or not).

I want to be **very clear** here: Using one approach or another doesn't make you better, harder, faster, stronger, smarter, or cooler. I strongly encourage everyone to use the tools that both make sense to them and will work for the task they have at hand. For this demonstration, I needed to pick something, so I picked this set of tools. What you should also realize (and which I'll bring up throughout this) is that each tool *does* come with a set of power relations embedded within it, a set of choices as to how it's intended to be used, and who it's intended to be used by. We'll splend plenty of time attempting to subvert some of these expectations, that doesn't mean we'll succeed. 

## Finally getting started

Alright, so at this point I'm going to assume that you have the notebook file downloaded and that you have Anaconda installed (or are familiar enough with another system to have chosen to use that instead).

**Please note: Most of these instructions are written for Windows. Linux is pretty straightforward, but Mac OS can sometimes through a few curveballs. Check out [here](https://docs.anaconda.com/free/anaconda/install/mac-os/) on that front**. 

What we need to do now is:
1. Create an environment for this workshop
2. Install the necessary libraries
3. Launch Jupyter
4. Open this notebook from within Jupyter to make it fully interactive

### Step 1: Creating an environment
I'm going to do this using the Anaconda Prompt and the command line. There are many other ways, but I enjoy command line interfaces as noted above. 

First, open up your 'Anaconda Command Prompt' program (again, assuming you are in Windows. In Linux, you can usually just drop into your bash shell or whatever you use so long as your pathing is set up correctly, same with Mac and terminal).

Type `conda create -n digiws python=3.8` and hit enter. Then confirm when prompted.
This is going to create an [anaconda environment](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/environments.html) for everything you do with this project. You generally want to create environments for all your major projects as that will keep version conflicts between libraries to a minimum. In other words, if one of your projects needs version 3.1 of GeoPandas to work and another needs version 3.2, if you tried to work in the same environment... you'd get errors. By setting up environments for each project, you make sure that version conflicts are minimized (note, I did not say eliminated...). It's also going to install python version 3.8 into that environment. You can choose another one, I just picked 3.8 because when I originally wrote some of what comes later in this lab, that's what I happened to be using.

Once that's done, you'll want to activate said environment. On Windows, you type `conda activate digiws`; on Mac/Linux, the command is `source activate digiws`. 

Before your drive prompt, you should now see a little `(digiws)` that lets you know which environment you have active. If you ever forget **any of this** (as I often do), the [Conda Cheat Sheet](https://docs.conda.io/projects/conda/en/4.6.0/_downloads/52a95608c49671267e40c689e0bc00ca/conda-cheatsheet.pdf) is a lifesaver.

### Step 2: Installing the libraries

We're going to use a few libraries today just to set up a sample workflow from scraping data, to geocoding it, to putting it on a very simple webmap. I know that's not super exciting if you've done this stuff before, but I think there are interesting questions at each step that make working through them worthwhile. I'm also going to show you how you can loop ChatGPT into your workflow if you want to.

The libraries we'll use are: `jupyter, geopandas, geopy, plotly_express, and overpy`
To install them, you type `conda install -c conda-forge jupyter geopandas geopy plotly_express overpy`
You could leave out the `-c conda-forge` for many packages, it sets the 'channel' Anaconda looks for the packages in as conda-forge; here, we need it for GeoPy.

To explain each package: jupyter is jupyter notebooks; geopandas is an [open-source GIS](https://geopandas.org/en/stable/); geopy is wrapper for various [geocoders](https://geopy.readthedocs.io/en/stable/index.html); Overpy is an [overpass turbo wrapper](https://pypi.org/project/overpy/) that lets us query OSM data; and Plotly Express is a [high-level API for Plotly](https://plotly.com/python/plotly-express/) that will help us make interactive maps, basically.

What these libraries will allow us to do scrape some web data, process it a bit, then visualize it on an interactive map. 
While they install, sit back and feel like you're the star of a Hackers reboot.


### Step 3: Launch Jupyter
Once everything is nice and installed, type in:
`jupyter notebook`

This will open up the Jupyter navigation pane in your default web browser. I use Edge because it's so bad it makes me laugh. I recommend against this.

### Step 4: Open this notebook... again

Navigate to wherever you have this notebook saved using the Jupyter navigator, and double click to open this notebook. It will open in a new window and if you click on any of the cells, you'll be able to interact with them.



## Our goal: scrape some public data, geocode it, and put it on a map
_(then, at the end, I'll show you how to do something similar with ChatGPT)_

#### The following cell is where the code begins.
First, I will import the geopandas library.
Then, I will use a very simple loop to iterate across an API and grab all of the data it spits back.
I am going to use Tacoma city license data here, because I have been based out of the wonderful city of Tacoma for the past nine years. 

In [8]:
# This will take a few minutes to run. It relies on geopandas *and* pandas (installed with geopandas)
# You'll see why you need both below

import geopandas, pandas

# In python, beginning a line with # will mark it as documentation and not to be executed.
''' You can also do that for multiple lines
    like this.
    But, since this is really just a multi-line string, make sure you end it or you will get an error! '''

# The following code is a very basic way to iterate across results to build up data through an api
# This is a very common and not particularly elegant. Much of the code here today won't be. That's ok!

# Please note, I am 'cheating' in that I KNOW how many results are in the table by looking at the metadata here:
# https://geohub.cityoftacoma.org/datasets/tacoma::tacoma-business-license/about

# Since I KNOW there are 29k entries, I am going to stop my loop once I have that many.
# There are ways to do this WITHOUT knowing the total number of entries.
# One common approach there is to automatically 'break' your loop when it runs out of data (results less than 1k here)


# The API will only give me 1,000 results at a time (many APIs have limits like this)
# Sometimes, the API will be documented (as in this case)
# Othertimes... experimentation and a willingness to... tactfully ignore terms of service will see you through

# I am going to set up a loop that will pull chunked results of 1k sequentially
# For the purposes of this tutorial, I am *not* error checking *nor* checking for duplicates.
# This is a decision that I'm knowingly making, but it will impact my results.
# These sorts of decisions suffuse code.


x = 0
while x < 30000:
    # I'm constructing a url here that has a set limit and then a stepped offset.
    # In other words, I'm building my query into the html - a pretty standard approach for a REST Service
    # Wikipedia actually has a great introduction to REST: https://en.wikipedia.org/wiki/Representational_state_transfer
    # Someday, I'll write my long-overdue 'REST as Weberian Bureaucracy piece'
    # Please be in touch if you'd like to collaborate on it. ;)
    
    url = 'https://services3.arcgis.com/SCwJH1pD8WSn5T5y/arcgis/rest/services/TACOMA_BUSINESS_LICENSE/FeatureServer/0/query?outFields=*&where=1%3D1&f=geojson&&resultOffset='+str(x)
    if x == 0:
       
        # if this is the first time through, create the geodataframe and call it df
        # there are ways to not include this step, but when my code is not time constrained, I like to be explicit
        
        df = geopandas.read_file(url)
        
    else:
        # if the dataframe exists, add to it.
        #I know how to do this by READING GEOPANDAS DOCUMENTATION here: https://geopandas.org/mergingdata.html
                
        df2 = geopandas.read_file(url)
        
        # In 'ye olden days', you could use a geopandas .append() method
        # Pandas deprecated that and geopandas hasn't created its own again.
        # Instead, it's easiest (and even recommended) to just use the pandas concat function, as so:
        df = pandas.concat([df, df2])
        
        # Libraries build on libraries, networks build on networks, infrastructure builds on infrastructure, etc.
        
    # Now let's add to our offset!
    x = x + 1000

# Now that we're done, let's check how many entries we have *and* see what the data looks like.

print(df.head())
print(len(df.index))

   LICENSE_NUMBER                        BUSINESS_NAME  \
0     500012761.0                   NISSAN-INFINITI LT   
1     500081953.0        APOLLO MECHANICAL CONTRACTORS   
2     500032593.0                KALLES PROPERTIES INC   
3     500009599.0  NISSAN MOTOR ACCEPTANCE COMPANY LLC   
4     500045940.0                           ROBIN JUPP   

                            OWNER_NAME HOUSE_NUMBER_AND_STREET_NAME  \
0                   NISSAN-INFINITI LT                 1 NISSAN WAY   
1                 APOLLO ENVIRONMENTAL           1207 W COLUMBIA DR   
2           KALLES PROPERTY MANAGEMENT                  2702 E MAIN   
3  NISSAN MOTOR ACCEPTANCE COMPANY LLC                 1 NISSAN WAY   
4                           ROBIN JUPP               817 30TH AVE S   

  UNIT_NUMBER P_O__BOX       CITY STATE    ZIP_CODE  NAICS_CODE  \
0         NaN      NaN   FRANKLIN    TN       37067    532112.0   
1         NaN      NaN  KENNEWICK    WA  99336-3459    238390.0   
2         # B      NaN 

### A few things you might notice

First off, the official documentation for this data set ([here](https://geohub.cityoftacoma.org/datasets/tacoma::tacoma-business-license/about)) says there are 28,539 entries; yet, we have a database with 28,594. Not a huge difference, but just another space in which automated processes can - due to negligence, intention, or just lack of familiarity - introduce errors. What's actually correct? Who collected the data? Etc.

Second, see that geometry column at the end and how it has `None` as its entry for each. That's awesome. The *reason* for this is an error in how the ESRI API the city of Tacoma uses is configured. 

I **know** they're aware of this issue as I've spoken to them about it for the past two years. It's not getting fixed. It's not a priority and I've been told I'm the only person that's noticed it. Incidentally, the previous API (Socrata) did not have this error, but cost more in upkeep. Capitalism and public funding rear their heads once more!

**But**, if we poke around the dataset, we also see that many of them contain data in columns `HOUSE_NUMBER_AND_STREET_NAME`, `CITY`, and `STATE` (`ZIP_CODE`) as well. We're going to use GeoPy to construct addresses and geocode this information.

This sort of futzing and adaptability is extremely useful when working with datasets. Never give up!

In the next cell, I'm going to create a very rudimentary geocoder using [Nominatim](https://nominatim.org/), I'll then discuss the limitations and give you suggestions on how to improve it.


In [6]:
from geopy import geocoders

#This is going to create a function that takes in one parameter and then geocodes it

def geo(location):
    ''' takes a location as a string and returns a lat and long pair'''
    # Note - You need to specify a 'user agent' so Nominatim can keep track of who you are
        g = geocoders.Nominatim(user_agent="DigiWS 1")
    
    # Now, let's feed into that geocoder the 'location' parameter that the function ingests
    loc = g.geocode(location)
    
    # Ok, let's have the function pass back the latitude and longitude as a pair
    return loc.latitude, loc.longitude

# One of these will work, one of these will not.
# Think about what this means for *error catching*

print(geo('906 8 Ave SW #226, Calgary'))
print(geo('zbjaewpito'))

(51.046371050000005, -114.08214283920702)


AttributeError: 'NoneType' object has no attribute 'latitude'

### A wild error appears!

The first result is (roughly) your present location. The second result, some nonsense string I through into the geocoder, throws an error - if the geocoder can't geocode the result, it can't return something.

Here's where you get your first chance to really work on something yourselves.

I encourage you to work in pairs and to ask questions! Especially if this is your first time working in this environment, there's so much being covered so quickly. Don't feel bad if you're a bit confused and **do** ask questions. I promise you, it took me *years* to get comfortable writing code; no one expects you to do it in an hour. :)

Here are two ideas though:
1. Use an if/then statement to check that loc exists, and - if it does not - return None, None (or something else)
2. Even fancier, if it doesn't exist (or if you get any error, really, such as a rate limit), load up a different geocoder and try that.

The latter is the more robust solution that helps stay within terms of service; however, the first one is sufficient for this exercise - again, this is a *decision* you all (and I) are making; it's not inevitable. If you do want to pursue it, the GeoPy documentation covers the various geocoders [available](https://geopy.readthedocs.io/en/stable/index.html#module-geopy.geocoders).


## Group question 1: Produce a geocoded table of the taxi business registered in Tacoma

If you get bored or stumped or short on time, I'm going to give you the solution two cells down.


In [ ]:
from geopy import geocoders

def geo(location):
    ''' takes a location as a string and returns a lat and long pair'''
    # Your cool stuff goes here

print(geo('906 8 Ave SW #226, Calgary'))
print(geo('zbjaewpito'))

# Try your best working together!

.

.

.

.

.

.

.

.

.

.

.

.

.

.

.

.

.


And remember my solution really isn't a great one.

In [11]:
from geopy import geocoders

def geo(location):
    ''' takes a location as a string and returns a lat and long pair'''
    # Note - You need to specify a 'user agent' so Nominatim can keep track of who you are
    g = geocoders.Nominatim(user_agent="DigiWS 2", timeout=10)
    # I'm setting a timeout that's larger here in case a bunch of people hit this at once
    
    loc = g.geocode(location)
    if loc is not None:
        return loc.latitude, loc.longitude
    else:
        return None, None
    
print(geo('1900 Commerce Street, Tacoma, WA'))
print(geo('zbjaewpito'))

(47.245217499999995, -122.43981965250066)
(None, None)


### Putting it all together and getting it on a map (almost)

One thing you can notice about the above dataset is that it contains a lot of entries that don't have addresses to geocode. You'll also notice, if you've been reading documentation, that Nominatim really isn't meant for super high loads. For the sake of speed and being kind to an open-source, free product, I'm going to first select only Taxi Services from our dataset *and then* drop all of the ones that don't contain addresses to geocode.

While not a focus of this tutorial, dropping data you know won't work **before** you hit the API (which often costs money) can be both a cost and time-saving technique.


In the following cell, I'll first use a query to define only Taxi Services and then use .dropna() (a geopandas method you can read about [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) which allows me to remove data without the values I want in them).

#### This next cell will take *awhile* to run. Take a break, chat about questions you have, etc. 
It's because we're using a free geocoder and it's not meant to handle bulk loads. But, it'll do for today.



In [27]:
# I'm only interested in the taxi licenses AND I've read the Nominatim terms of service about limiting bulk queries
# Why don't I filter out my data frame to only include taxi licenses and then geocode those?
# Hmmm... how could I do that? Perhaps the NAICS code?
# I can read about pandas QUERIES here: https://www.sharpsightlabs.com/blog/pandas-query/ (or the formal documentation)

taxi_df = df.query('NAICS_CODE == 485310.0')

# A lot of these don't have data in them, so let's drop those as well. I'll use .dropna()
# We use the .copy() at the end to avoid setting values on a slice
# Where could you read about that, do you think?
taxi2 = taxi_df.dropna(axis=0, how='any', subset=['HOUSE_NUMBER_AND_STREET_NAME']).copy()

# This next step is unnecessary as I could construct this information each time I need to
# However, I'm going to create a new column in my dataframe that has the address location I want to pass to my geocoder
# The format will read "# Street Name, City, State"
taxi2['location'] = taxi2['HOUSE_NUMBER_AND_STREET_NAME'] + ', ' + taxi2['CITY'] + ', ' + taxi2['STATE']

# I'm also going to create empty columns called lat and long.
# This is where my solution begins to veer from the most elegant one, see if you could condense these steps
# Or, better yet, do a wholly different approach. This approach will get slower for extremely large datasets.
# You can see some of the timings of how these approaches compare here: 
# https://medium.com/swlh/how-to-efficiently-loop-through-pandas-dataframe-660e4660125d

taxi2['lat'] = 0
taxi2['long'] = 0

for i in range(len(taxi2)):
    taxi2.loc[i, 'lat'], taxi2.loc[i, 'long'] = geo(str(taxi2['location'].iloc[i]))
    
#Let's tidy things up once more to make a 'clean' geodataframe with only addressable results...
taxi3 = taxi2.dropna(axis=0, how='any', subset=['lat']).copy()

print(taxi4.head())

     LICENSE_NUMBER           BUSINESS_NAME  OWNER_NAME  \
590     500086620.0     HAROLD R LETOURNEAU  TNC DRIVER   
765     500113564.0          LATOYA D LEWIS  TNC DRIVER   
974     500191202.0  JONATHAN DE LOS SANTOS  TNC DRIVER   
3       500189887.0       ABDULKADIR GESHOW  TNC DRIVER   
140     500145287.0          DMITRI FEDOROV  TNC DRIVER   

    HOUSE_NUMBER_AND_STREET_NAME UNIT_NUMBER P_O__BOX      CITY STATE  \
590           35209 172ND AVE SE         NaN      NaN    AUBURN    WA   
765        18808 109TH AVE CRT E         NaN      NaN  PUYALLUP    WA   
974             15819 42ND AVE S         NaN      NaN   TUKWILA    WA   
3           18808 105TH AVE CT E         NaN      NaN  PUYALLUP    WA   
140             5132 BRIGHTON ST         NaN      NaN   SEATTLE    WA   

       ZIP_CODE  NAICS_CODE         NAICS_CODE_DESCRIPTION  \
590  98092-2616    485310.0  Taxi and Ridesharing Services   
765       98374    485310.0  Taxi and Ridesharing Services   
974  98188-2644    4

In [28]:
print(len(taxi4))

522



## Look at that!

We now have a dataset of 522 geolocated Taxi Services registered in the city of Tacoma. Of course, we started with _either_ 28,594 or 28,593 of _all_ businesses and these are, of course, only the businesses in the Tacoma City spatial data set, and... and... and...

The point isn't that this isn't rad. It is. It's that we're making a series of selections that shape our eventual visualization - from seen to unseen; our data is never inevitable. Deep.

Anyway, let's make a map using PyPlot Express.
Note a few things below: 
1. The color is meaningless (I could process the open date into something meaningful, I have not)
2. There are many other settings you can tinker with.
3. Check [here](https://plotly.com/python/plotly-express/#tile-maps)


In [35]:
import plotly.express as px

fig = px.scatter_mapbox(taxi4, lat="lat", 
                        lon="long", color="BUSINESS_OPEN_DATE", 
                        color_continuous_scale=px.colors.cyclical.IceFire, 
                        zoom=6, mapbox_style="carto-positron")
fig.show()

Alright, we've made a map. We've gone from ~29k data points to ~500. We've thrown up colors based on unprocessed data... etc.

We've made a lot of choices and that's ok. But, the basic framework is intact:
1. Set up an environment with the libraries you need.
2. Acquire some data and process it (geocode, reformat, etc.)
3. Visualize


But, all of that was rather finicky. Now, I want to show you how to use [ChatGPT](https://chat.openai.com/) to generate geodataframe data and visualize it. ChatGPT can be finicky (and sometimes, or even often, outright incorrect), but it can also be really handy if you want to get some data and code working quickly. Below are the intstructions I gave it (including a response when its original code produced an error).

Notice that I am using Open Street Map data here. ChatGPT can work with other data sets and you should explore which ones would make sense for you (could you get it to work with the Tacoma dataset? The answer is yes, but not by directly asking for it to produce a query for the Tacoma API, but by specifying the form of the API and then doing a bit of modifying yourself).


> Using python and overpy, can you write a query for overpass turbo that will return the location of all bars within the Boston city limits as a geopandas dataframe?

(resulting code returned an error, I copied the error into the interface)

> I'm getting the following error with the above code, can you fix it?

> ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 24
     22 # Extract the location data from the response
     23 data = []
---> 24 for element in response.elements:
     25     if element.type == "node":
     26         lon = element.lon

> AttributeError: 'Result' object has no attribute 'elements'


The following code was the result, try running it.




In [44]:
import overpy
import geopandas as gpd

# Create an Overpass API object
api = overpy.Overpass()

# Define the query for bars within Boston city limits
query = """
[out:json];
area[name="Boston"]->.a;
(
  node["amenity"="bar"](area.a);
  way["amenity"="bar"](area.a);
  relation["amenity"="bar"](area.a);
);
out center;
"""

# Send the query to Overpass API
result = api.query(query)

# Process the query result and extract coordinates
coordinates = []
for element in result.nodes + result.ways + result.relations:
    if hasattr(element, 'lat') and hasattr(element, 'lon'):
        lat = element.lat
        lon = element.lon
        coordinates.append((lat, lon))
    elif hasattr(element, 'center'):
        lat = element.center.lat
        lon = element.center.lon
        coordinates.append((lat, lon))

# Create a GeoPandas DataFrame from the coordinates
df = gpd.GeoDataFrame(geometry=gpd.points_from_xy([c[1] for c in coordinates], [c[0] for c in coordinates]))

# Set the coordinate reference system (CRS) to WGS84 (EPSG:4326)
df.crs = "EPSG:4326"

# Print the resulting DataFrame
print(df)



                      geometry
0   POINT (-71.06529 42.34364)
1   POINT (-71.07530 42.34223)
2   POINT (-71.04330 42.34616)
3   POINT (-71.04316 42.34627)
4   POINT (-71.09671 42.34704)
..                         ...
92  POINT (-71.05967 42.35751)
93  POINT (-71.05562 42.35431)
94  POINT (-71.06462 42.35152)
95  POINT (-71.06466 42.35135)
96  POINT (-71.06476 42.35090)

[97 rows x 1 columns]


That's cool and all, but it is solely the point information. Let's ask it to split out the lat and lon data (not necessary, but makes pyplot a touch easier) and to add the bars' names...

> Please add columns for latitude and longitude separately and add the bars' names to the dataset as string values

This produces a ValueError where it had defined the 'name' to be a string, I asked it to correct that and got the following code after a bit of back and forth (please note, I could have corrected the error myself; but, the point is that I did not need to):

In [45]:
import overpy
import geopandas as gpd

# Create an Overpass API object
api = overpy.Overpass()

# Define the query for bars within Boston city limits
query = """
[out:json];
area[name="Boston"]->.a;
(
  node["amenity"="bar"](area.a);
  way["amenity"="bar"](area.a);
  relation["amenity"="bar"](area.a);
);
out center;
"""

# Send the query to Overpass API
result = api.query(query)

# Process the query result and extract coordinates and names
data = []
for element in result.nodes + result.ways + result.relations:
    if hasattr(element, 'tags') and 'name' in element.tags:
        name = element.tags['name']
    else:
        name = 'Unnamed Bar'
        
    if hasattr(element, 'lat') and hasattr(element, 'lon'):
        lat = element.lat
        lon = element.lon
        data.append({'Name': name, 'Latitude': lat, 'Longitude': lon})
    elif hasattr(element, 'center'):
        lat = element.center.lat
        lon = element.center.lon
        data.append({'Name': name, 'Latitude': lat, 'Longitude': lon})

# Create a GeoPandas DataFrame from the data
df = gpd.GeoDataFrame(data, geometry=gpd.points_from_xy([d['Longitude'] for d in data], [d['Latitude'] for d in data]))

# Set the coordinate reference system (CRS) to WGS84 (EPSG:4326)
df.crs = "EPSG:4326"

# Print the resulting DataFrame
print(df)




                 Name    Latitude    Longitude                    geometry
0   J.J. Foley's Cafe  42.3436416  -71.0652939  POINT (-71.06529 42.34364)
1        Sister Sorel  42.3422326  -71.0752954  POINT (-71.07530 42.34223)
2           Birch Bar  42.3461583  -71.0433047  POINT (-71.04330 42.34616)
3            City Bar  42.3462734  -71.0431626  POINT (-71.04316 42.34627)
4        Bleacher Bar  42.3470437  -71.0967137  POINT (-71.09671 42.34704)
..                ...         ...          ...                         ...
92        haley.henry  42.3575054  -71.0596738  POINT (-71.05967 42.35751)
93        Bubble Bath  42.3543113  -71.0556163  POINT (-71.05562 42.35431)
94            The Tam  42.3515157  -71.0646200  POINT (-71.06462 42.35152)
95           4th Wall  42.3513531  -71.0646629  POINT (-71.06466 42.35135)
96           Bar Moxy  42.3509022  -71.0647555  POINT (-71.06476 42.35090)

[97 rows x 4 columns]


Now, I'm going to ask it to visualize the results as a mapbox scatterplot using plotly express, at zoom 8, and centered on boston:

> Can you visualize the resulting data as a mapbox-tile scatterplot using pyplot express at zoom 8 and centered on Boston?


In [48]:
import overpy
import geopandas as gpd
import plotly.express as px

# Create an Overpass API object
api = overpy.Overpass()

# Define the query for bars within Boston city limits
query = """
[out:json];
area[name="Boston"]->.a;
(
  node["amenity"="bar"](area.a);
  way["amenity"="bar"](area.a);
  relation["amenity"="bar"](area.a);
);
out center;
"""

# Send the query to Overpass API
result = api.query(query)

# Process the query result and extract coordinates and names
data = []
for element in result.nodes + result.ways + result.relations:
    if hasattr(element, 'tags') and 'name' in element.tags:
        name = element.tags['name']
    else:
        name = 'Unnamed Bar'
        
    if hasattr(element, 'lat') and hasattr(element, 'lon'):
        lat = element.lat
        lon = element.lon
        data.append({'Name': name, 'Latitude': lat, 'Longitude': lon})

# Create a GeoPandas DataFrame from the data
gdf = gpd.GeoDataFrame(data, geometry=gpd.points_from_xy([d['Longitude'] for d in data], [d['Latitude'] for d in data]))

# Set the coordinate reference system (CRS) to WGS84 (EPSG:4326)
gdf.crs = "EPSG:4326"

# Define the center coordinates for Boston
boston_center = {'lat': 42.3601, 'lon': -71.0589}

# Plot the scatterplot using plotly express with Mapbox tiles
fig = px.scatter_mapbox(gdf, lat='Latitude', lon='Longitude', hover_name='Name', zoom=12, center=boston_center)
fig.update_layout(mapbox_style='open-street-map')
fig.show()


## Ok, so now what?

Well, now I want you all to think about a few things:
1. Do you understand the code written by ChatGPT? Is it how you would write code? Did you understand the code that we walked through manually? Does the difference matter?
2. Could you keep interacting with it to produce more and better visualizations and analysis (ask it to add different tiles, to bring in additional data, etc.)?
3. What decisions is ChatGPT making? How does it elide choices we walked through (or does it)? What power relations does it normalize?

There's a lot more to ask. There's a *ton* more to think through and potentially do. I don't have answers to any of the above; but, I would like to make clear a few things. I absolutely do not believe everyone needs to code or even interact with code, that's not what being a 'digital geographer' is about and I hope this workshop didn't give you that idea! I *do* think that thinking through what code is and what it does, how it shapes what we see and don't, how queries reflect data and how data (doesn't) reflect reality are fundamental questions.

There's space for so many approaches, but here - at this tutorial - I hope I got you thinking about the decisions we make when we write code and what it means, a bit, to think with a corporate AI. What it enables and what it constrains.

Cheers,

Jim Thatcher

Oregon State University
